save-osv/src/main/kotlin/com/saveourtool/save/osv/service/OsvService.kt

```diff
@@ -0,0 +1,137 @@
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.toMono
+
+import java.io.InputStream
+
+import kotlinx.datetime.LocalDateTime
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.decodeFromStream
+
+/**
+ * Service for vulnerabilities
+ */
+@Service
+class OsvService(
+    private val osvStorage: OsvStorage,
+    private val vulnerabilityService: IVulnerabilityService,
+) {
+    private val json = Json {
+        prettyPrint = false
+    }
+
+    /**
+     * Decodes [inputStream] and saves the result
+     *
+     * @param inputStream
+     * @param authentication who uploads [inputStream]
+     * @return save's vulnerability names
+     */
+    fun decodeAndSave(
+        inputStream: InputStream,
+        authentication: Authentication,
+    ): Flux<String> = { json.decodeSingleOrArrayFromStream<RawOsvSchema>(inputStream) }
+        .toMono()
+        .flatMapIterable { it }
+        .flatMap {
+            save(it, authentication)
+        }
+
+    /**
+     * Decodes [content] and saves the result
+     *
+     * @param content
+     * @param authentication who uploads [content]
+     * @return save's vulnerability names
+     */
+    fun decodeAndSave(
+        content: String,
+        authentication: Authentication,
+    ): Flux<String> = { json.decodeSingleOrArrayFromString<RawOsvSchema>(content) }
+        .toMono()
+        .flatMapIterable { it }
+        .flatMap {
+            save(it, authentication)
+        }
+
+    /**
+     * Saves [vulnerability] in S3 storage and creates a new entity in save database
+     *
+     * @param vulnerability
+     * @param authentication who uploads [vulnerability]
+     * @return save's vulnerability name
+     */
+    private fun save(
+        vulnerability: RawOsvSchema,
+        authentication: Authentication,
+    ): Mono<String> = osvStorage.upload(vulnerability).blockingMap {
+        vulnerabilityService.save(vulnerability.toSaveVulnerabilityDto(), authentication).name
+    }
+
+    /**
+     * @param name [VulnerabilityDto.name]
+     * @return found OSV
+     */
+    @OptIn(ExperimentalSerializationApi::class)
+    fun findBySaveName(
+        name: String,
+    ): Mono<RawOsvSchema> = blockingToMono {
+        vulnerabilityService.findByName(name)
+    }
+        .switchIfEmptyToNotFound {
+            "Not found vulnerability $name"
+        }
+        .map { vulnerability ->
+            vulnerability.vulnerabilityIdentifier.orNotFound {
+                "Vulnerability $name doesn't have vulnerabilityIdentifier"
+            }
+        }
+        .flatMap { id ->
+            osvStorage.downloadLatest(id)
+                .collectToInputStream()
+                .map { json.decodeFromStream<RawOsvSchema>(it) }
+        }
+
+    companion object {
+        private fun RawOsvSchema.toSaveVulnerabilityDto(): VulnerabilityDto = VulnerabilityDto(
```

Comment:
I quickly looked at this PR and cannot see adapter where you need to convert RAW -> DTO, where is it?

Reply:
Currently there is no a dedicated service for it or any engine. Now it's just a private function which creates `VulnerabilityDto` from `RawOsvSchema`. It's this method

Reply:
introduced OsvProcessor with default implementation now.
In future, we can detect OsvProcessor by ID and finish parsing of custom fields and then convert to VulnerabilityDto.